### ISSUE TITLE
libbpf as a main part of program loading

### ISSUE THREAD

**Issue by BurntBrunch** [iovisor/bpftrace#2334]
This is a first draft of the refactoring steps discussed in #2322.

The overall goal is to delegate more of the work to libbpf and gradually stop dealing with details that libbpf can already handle.

## Program and map loading

1. Emit in-memory ELF and build the BpfBytecode map out of it via ELF
1.1. This step remove the ORC usage but keeps the current bag-of-sections approach.
3. Emit global variables for all the maps in the libbpf format, which requires BTF. As I found out in #2321, this requires LLVM 10.
4. Import libbpf 0.8.1 as a git submodule
  3.1 Configure local builds to use it by default.
  3.2 Create PACKAGING.md and document the minimum LLVM and libbpf versions. 
    3.2.1 Optionally document every single llvm dependency (individual library names)
5. Use `bpf_object__open_mem` to iterate over the programs but pass the instructions directly to the syscall
  4.1. Remove the non-`bpf_prog_load` paths around this time
  4.2. Ensure we call `bpf_program__set_autoattach(false)` on all programs
  4.3. If necessary (?), call `bpf_program__set_expected_attach_type`
6. Be able to cleanly `bpf_object__load` the programs and maps

At this point, program relocation and loading happens with libbpf and we can use `bpf_program__fd` to pass the resulting fd into the existing attachment infrastructure.
An interesting angle here is that an external program can at this point easily load the output of `--emit-elf`. This might be interesting for AOT though I haven't gone through that code yet.

## Program attachment

From here on out, the path is roughly:
  * For each program type:
    * Use libbpf to actually attach the program via `bpf_program__attach*`
    * (Optionally) Migrate to the libbpf naming convention for that attachment type and use autoattach
  * In terms of new features, at this point we can potentially use the ksyscall logic to automatically attach to the right architecture-specific syscall handler via kfunc (instead of tracepoints).

### USDT suport

The biggest question mark in my mind is the USDT support in libbpf. While libbpf has helpers we can't really use for access to usdt params in BPF, it also requires a ["usdt manager" to be active](https://github.com/libbpf/libbpf/blob/master/src/usdt.c#L24).

Overall, this area needs more investigation - it might be the case that bpftrace itself would embed this BPF-side logic and load it when it encounters usdt probes. Or, it might be the case that bpftrace doesn't fit into this model of working with USDT.

**Comment by anakryiko** [comment]
> pass the instructions directly to the syscall

and 

> program relocation and loading happens with libbpf 

are in diagreement. Which one it is? I think, if at all possible, leave loading BPF programs to libbpf and `bpf_object__load()`. Load step performs transformations, CO-RE relocations. Instructions you get with `bpf_program__insns()` before load happened is not a final instructions, you generally can't load them into kernel with `bpf_prog_load()`. With `bpf_program__set_autoload()` you have control what should and shouldn't be loaded into kernel, so bpftrace should have enough flexibility.

There is also custom `SEC()` support in libbpf that would allow to intercept BPF program load, but I'd leave it as a last resort.

**Comment by BurntBrunch** [comment]
> Which one is it?

It's a bit ambiguously worded. I meant it more as "use bpf_object to discover the maps and programs but do relocation and loading the way it's done today." 

There may not be much value in that step other than separating the refactoring of ownership and internal bookkeeping of probe to program mappings, tbh. There are, however, a lot of relationships to clean up and clarify in the code (e.g wildcard expansions into one or multiple programs depending on the contents of the program) and this step can make them more obvious.

At the end of the day, libbpf should do the loading, it's just a question of how we get there.

**Comment by anakryiko** [comment]
> libbpf should do the loading, it's just a question of how we get there

yeah, I was trying to understand if that's the plan, thanks! It makes sense to go through intermediate steps to get there, yep.

**Comment by BurntBrunch** [comment]
#2321 got a little stuck on dealing with old libbpf versions, so I decided to hurry up and do the import into the repo. Uhm, that hasn't gone according to plan.

https://github.com/BurntBrunch/bpftrace/commits/vendor-libbpf has my work in progress. I've opted to import as a submodule, so it's easier to navigate, but I'm open to just using ExternalProject and letting cmake check it out.

The problem is that cmake is ... limited. There are two main problems:

1) Symbol checking (HAVE_BCC_X) works at *generation time*. This means it cannot depend on anything internal to the cmake build, like an ExternalProject.
2) Link order between external libraries and internal libraries cannot be customized. In particular, we need the external libbcc.a to come _before_ the internal libbpf.a when STATIC_LINKING=ON, but cmake doesn't seem to be able to express this. Maybe IMPORTED targets for all the external libraries would help but I don't have the energy to try.

Given all this, I see two approaches:
1) Build libbpf outside of the cmake build, similar to how CI does it. This is easiest.
2) Keep libbpf inside the cmake build but pull in libbcc too so we can express the dependencies. This is not my preferred path, given the amount of work.

@viktormalik @danobi thoughts?

**Comment by viktormalik** [comment]
I've picked up this work and now have a working implementation. It's a lot of code so I'll try to split it into smaller parts to simplify reviewing.

Here's the list of planned changes:
1. Fix generated LLVM to adhere to what libbpf expects (fix function names, add license, ...).
2. Rework probe expansion to do it in LLVM IR. See #3005 for more details on this topic.
3. Load all BPF programs at once from `BpfBytecode`.
4. Fully delegate program loading (and map creation) to libbpf (via `bpf_object`).

**Comment by anakryiko** [comment]
@viktormalik please make sure that .BTF.ext is also incorporated (it could be step 5): function info, line info, and especially CO-RE relocations are super important for modern-day BPF programs.

**Comment by viktormalik** [comment]
> I've picked up this work and now have a working implementation. It's a lot of code so I'll try to split it into smaller parts to simplify reviewing.
> 
> Here's the list of planned changes:
> 
> * [x]  1.  [Make codegen compatible with libbpf #3153](https://github.com/bpftrace/bpftrace/pull/3153)
> * [x]  2. Rework probe expansion to do it in LLVM IR. See [[RFC] Probe expansion in codegen #3005](https://github.com/bpftrace/bpftrace/issues/3005) for more details on this topic.
> * [ ]  3. Load all BPF programs at once from `BpfBytecode`.
> * [ ]  4. Fully delegate program loading (and map creation) to libbpf (via `bpf_object`).
